### ISSUE TITLE
include/grpcpp/test/mock_stream.h is not installed by `make install`

### ISSUE THREAD

**Issue by sandeepgd** [grpc/grpc#20208]
# What version of gRPC and what language are you using?
v1.20.1 & c++

### What operating system (Linux, Windows,...) and version?
Ubuntu 16.04.6 LTS

### What runtime / compiler are you using (e.g. python version or version of gcc)
gcc version 5.4.0

### What did you do?
`cd /tmp && git clone -q --single-branch --depth 1 --branch v1.20.1 https://github.com/grpc/grpc.git && cd grpc && git submodule update --init && make CFLAGS='-Wimplicit-fallthrough=0' CXXFLAGS='-Wimplicit-fallthrough=0' > /dev/null && make install STRIP=/bin/true`

### What did you expect to see?
Expect following header file to be present:
`/usr/local/include/grpcpp/test/mock_stream.h`

### What did you see instead?
All other include files/directories are available except the `/usr/local/include/grpcpp/test` directory.

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

### Anything else we should know about your project / environment?
I need `mock_stream.h` to be able to write mock tests for my grpc service.

**Comment by nanahpang** [comment]
I don't understand your requirement here. So you want to write mock test but cannot use # include "mock_stream.h"? Is that what you mean?

**Comment by sandeepgd** [comment]
@nanahpang that's right, mock_stream.h is not installed as a public header, so I cannot include it. Sorry for the confusion.

**Comment by nanahpang** [comment]
Yes, the header  files under /test folder are not public headers. They are for test only. You need to add dependency in your build file while using them.

**Comment by sandeepgd** [comment]
@nanahpang I use `grpc` for my application but build `grpc` before I build my application. I do not build `grpc` along with my application. So, I cannot add dependency on specific `grpc` build targets in my application build file. Is there anyway to emit this file when building `grpc` by itself?

**Comment by sandeepgd** [comment]
@nanahpang Did you get a chance to look into this? Would greatly appreciate if this issue can be resolved soon. Thanks!

**Comment by nanahpang** [comment]
Hi, sorry for the late reply, I am busy with other things last week. 
The header files in test folder are private. I don't see a way you could use them outside grpc. But you could write your own test service based on that.

**Comment by sandeepgd** [comment]
Oh. Would grpc team consider making it public?
This public documentation describes how to unit-test client side logic:
https://grpc.github.io/grpc/cpp/md_doc_unit_testing.html
This uses `MockClientReaderWriter` but this class is defined in `mock_stream.h`. So I wonder if `mock_stream.h` was supposed to be public but is not due to a bug in the Makefile.

**Comment by nanahpang** [comment]
I need to talk with some folks in the team before responding to you. But if you are hurried to implement this, I would suggest you to write your own test service first.

**Comment by nanahpang** [comment]
Thanks for the document you sent, I will look into that.

**Comment by sandeepgd** [comment]
Thanks @nanahpang This is not urgent, I'll wait.

**Comment by nanahpang** [comment]
Hi, @sandeepgd 
After discussing this issue within the team, we think it is reasonable to release the headers in include/grpcpp/test folder. So I have opened a pull request for it. Thanks for sharing the information here! 
Good luck with your works.

**Comment by sandeepgd** [comment]
Awesome, thanks @nanahpang! Appreciate the quick fix. Please feel free to close the issue now.

**Comment by nanahpang** [comment]
Hi, @sandeepgd 
Could you check out the pull request locally, and see if it solve your problem? I have tested it myself, but want to make sure I could close this issue after merge it.

**Comment by nanahpang** [comment]
Friendly ping, @sandeepgd I have merged the pull request. Could you verify that your problem is solved? If not, you could reopen this issue.

**Comment by sandeepgd** [comment]
Hi @nanahpang, just verified that the test files are getting installed now. Thanks for the fix!

**Comment by nanahpang** [comment]
Previous PR is reverted, because it broke the build process in MacOS and Windows. Reopen this issue and investigate on another solution.

**Comment by Jumziey** [comment]
Any progress on this issue?

**Comment by sandeepgd** [comment]
It would be great to get this issue fixed again.

**Comment by jtattermusch** [comment]
See here https://github.com/grpc/grpc/issues/22937#issuecomment-630128278 
Some updates:
- note that installing with `make` (not cmake) is now deprecated
- with cmake you should be able to build/install grpc++_test (see the link for details).

**Comment by jtattermusch** [comment]
@sandeepgd  can you verify whether things got fixed (see my previous comment)?

**Comment by Dlougach** [comment]
@jtattermusch Can you please clarify, how is grpc++_test installable with CMake? I don't think that CMake currently installs any headers from `test/`

**Comment by jheaff1** [comment]
I too would very much like the mock_stream.h file available as part of the Cmake install

**Comment by pgeler** [comment]
same here, I think it would be beneficial if mock_stream.h is part of main installation

**Comment by GlebIgnatevEmlid** [comment]
+1, it's weird that you have to enable `gRPC_BUILD_TESTS` and build the tests in order to get the headers you need for your own tests.

**Comment by subox** [comment]
> +1, it's weird that you have to enable `gRPC_BUILD_TESTS` and build the tests in order to get the headers you need for your own tests.

Are there any plans to add exporting `mock_stream.h` without using of `gRPC_BUILD_TESTS`? Compiling with unnecessary `gRPC_BUILD_TESTS` increases the build time multiple times.

**Comment by misha-codethink** [comment]
Please consider installing `mock_stream.h` without `gRPC_BUILD_TESTS`: there are packages of `gprc` that do not install this header due to this configuration gate, at least
- Homebrew `grpc`, and
- OpenSUSE `grpc-devel`

**Comment by iMichka** [comment]
Homebrew maintainer here: I think this issue is for upstream to fix, we are not going to patch grpc to ship the file.

 guess that splitting the `grpc_cc_library` function call in two different calls could fix it (with testonly = True/False):
https://github.com/grpc/grpc/blob/6546fcd19640ead78749e135ab7514f2f5f93343/BUILD#L2775-L2797

Don't know if there is more to tweak, but maybe the fix is really easy to implement (I'll leave this to the grpc maintainers or anyone motivated enough to try this out and maybe submit a pull request here).

**Comment by glebignatieff** [comment]
+1 I have to patch the vcpkg port for this to work which is quite painful.
